groups/nts/ntsu/ntsu_socketutil.cpp

```diff
@@ -2361,46 +2525,39 @@ ntsa::Error SocketUtil::receive(ntsa::ReceiveContext*       context,
                                 const ntsa::ReceiveOptions& options,
                                 ntsa::Handle                socket)
 {
-    NTSCFG_WARNING_UNUSED(options);
-
     context->reset();
 
-    bsl::size_t capacity =
-        ntsa::MutableBuffer::totalSize(bufferArray, bufferCount);
-    if (capacity == 0) {
-        return ntsa::Error::invalid();
-    }
-
-    context->setBytesReceivable(capacity);
-
     const bool wantEndpoint = options.wantEndpoint();
+    const bool wantMetaData = options.wantMetaData();
+
+    msghdr msg;
+    bsl::memset(&msg, 0, sizeof msg);
 
     sockaddr_storage socketAddress;
     socklen_t        socketAddressSize;
 
     if (wantEndpoint) {
         SocketStorageUtil::initialize(&socketAddress, &socketAddressSize);
-    }
 
-    msghdr msg;
-    bsl::memset(&msg, 0, sizeof msg);
-
-    if (wantEndpoint) {
         msg.msg_name    = &socketAddress;
         msg.msg_namelen = socketAddressSize;
     }
 
+    ReceiveControl control;
+    if (wantMetaData) {
+        control.initialize(&msg);
+    }
+
+    bsl::size_t capacity =
+        ntsa::MutableBuffer::totalSize(bufferArray, bufferCount);
+    if (capacity == 0) {
```

Comment:
Here and and many other places this validation is now not the first thing `receive` function does. I think it is better to perform this check immediately and avoid unnecessray work done in a failure case.

Reply:
That is true, but I very much prefer the structure of the functions as they are written:
```
<common-setup>
<data-representation-specific-setup>
recvmsg
<common-cleanup>
```
We have a lot of overloads here and the functions as-written help me visually orient and detect that <common-setup> is correctly performed in each function. Calling `ntsu::SocketUtil::receive` with no capacity into which the copy the data is essentially an invalid call that I'm not particularly motivated by, given the previously described concern.